openmdao/solvers/ln_direct.py

```diff
@@ -42,23 +91,121 @@ def solve(self, rhs_mat, system, mode):
+
+                n_edge = len(rhs)
+                ident = np.eye(n_edge)
+
+                partials = np.empty((n_edge, n_edge))
+
+                for i in range(n_edge):
+                    partials[:, i] = self.mult(ident[:, i])
+
+            # Assemble the Jacobian
+            else:
+
+                # Must clear the jacobian if we switch modes
+                if self.mode != mode:
+                    self.setup(system)
+                self.mode = mode
+                sys_name = system.name + '.'
+                conn = system.connections
+
+                partials = self.jacobian
+                u_vec = system.unknowns
+
+                for sub in system.components(recurse=True, include_self=True):
+
+                    jac = sub._jacobian_cache
+
+                    # This method won't work on components where apply_linear
+                    # is overridden.
+                    if jac is None:
+                        msg = "The 'assemble' jacobian_method is not supported when " + \
+                             "'apply_linear' is used on a component (%s)." % sub.pathname
+                        raise RuntimeError(msg)
+
+                    sub_u = sub.unknowns
+                    sub_name = sub.pathname
+                    icache = self.icache
+
+                    for key in jac:
+                        o_var, i_var = key
+                        key2 = (sub_name, key)
+
+                        # Derivs wrt states, but states live in u_vec
+                        if i_var in sub.states:
+                            sub_p = sub_u
+                        else:
+                            sub_p = sub.params
+
+                        # Skip anything not relevant
+                        if o_var not in sub_u or i_var not in sub_p:
+                            continue
+
+                        # We cache the location of each variable in our jacobian
+                        if key2 not in icache:
+                            o_var_pro = sub_u.metadata(o_var)['top_promoted_name']
+                            meta = sub_p.metadata(i_var)
+                            i_var_pro = meta['top_promoted_name']
+
+                            # States are fine ...
+                            if i_var in sub.states:
+                                pass
+
+                            #... but inputs need to find their source.
+                            elif i_var_pro not in u_vec:
+                                if i_var_pro in conn:
+                                    i_var_src = conn[i_var_pro][0]
+                                else:
+                                    i_var_abs = meta['pathname']
+                                    i_var_src = conn[i_var_abs][0]
+
+                                # Promoted/Absolute gets kind of ridiculous
+                                # sometimes
+                                if i_var_src not in u_vec:
+                                    for name in u_vec:
```

Comment:
not necessary to loop here.  Since your outer loop is only over components and not groups, the names they use as jacobian keys will always be local names, so you can get the absolute name by  '.'.join((sub.pathname, jkey).  Given the abs name, you can get the top promoted name by looking it up in the to_prom_name dict from sub._probdata (_probdata refers to the same object in all systems, so that can be assigned to a temporary outside the loop).  To get the local promoted name for i_var_src, you can just look it up in system._sysdata.to_prom_name.